File: retail-mart-backend/retail-mart-backend/app/models/courier.py
"""
Registered courier / shipping partner entity.

Instead of free-text courier names on every shipment, Admin/Manager
users maintain a canonical list of courier partners here. The
ShipmentForm dropdown is populated from GET /api/couriers (active only).

Backward compatibility: Shipment.courier (string) is still the
serialized courier name used in all responses. When a shipment is
created with a courierId, we store both the FK and copy the name into
the string field — old shipments without a courier_id continue to
serialize correctly.
"""

from datetime import datetime

from app.extensions import db


class Courier(db.Model):
    """Registered delivery / courier partner."""

    __tablename__ = "couriers"

    id = db.Column(db.String(20), primary_key=True)
    name = db.Column(db.String(120), unique=True, nullable=False)
    code = db.Column(db.String(40), unique=True, nullable=False)          # e.g. "BLUEDART"
    contact_email = db.Column(db.String(150), nullable=True)
    # Template for building an external tracking link.
    # Use the literal placeholder {tracking_number} in the URL.
    # Example: "https://track.bluedart.com/?id={tracking_number}"
    tracking_url_template = db.Column(db.String(500), nullable=True)
    is_active = db.Column(db.Boolean, default=True, nullable=False)
    created_at = db.Column(db.DateTime, default=datetime.utcnow)
    updated_at = db.Column(db.DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)

    def to_dict(self) -> dict:
        """Matches src/types/courier.ts -> Courier"""
        return {
            "id": self.id,
            "name": self.name,
            "code": self.code,
            "contactEmail": self.contact_email,
            "trackingUrlTemplate": self.tracking_url_template,
            "isActive": self.is_active,
        }

File: retail-mart-frontend-connected/retail-mart-week2-final/src/types/courier.ts
/** Registered courier / shipping partner. Matches app/models/courier.py -> Courier.to_dict() */
export interface Courier {
  id: string;
  name: string;
  /** Uppercase alphanumeric code, e.g. "BLUEDART" */
  code: string;
  contactEmail: string | null;
  /**
   * External carrier URL template. Contains the literal placeholder
   * `{tracking_number}` which the backend replaces before returning
   * the final URL from GET /api/shipments/<id>/track.
   */
  trackingUrlTemplate: string | null;
  isActive: boolean;
}

export type CourierFormValues = Omit<Courier, "id">;
